### DISCUSSION TITLE
Beginner Issue: Can't Launch PiKaraoke

### DISCUSSION THREAD

**Discussion by StayAHDad** [vicwomg/pikaraoke#355] [Q&A] [answered]
Hello,

Please forgive my ignorance on this stuff - I am a complete amateur at all things GitHub, but am determined to set up a nice karaoke system for my wife. 

I've been trying to run PiKaraoke on my laptop (Windows), and have so far completed all the steps, including manually installing ffmpeg (as well as setting it in the user PATHS), as well as running the setup script to install python dependencies. All of this appears to have been installed correctly, although I have been googling like crazy in every step of the way, so I may have been mistaken here and there. However, when I try to finally launch the program and run "pikaraoke.bat" in my PiKaraoke folder, I get the following issue before everything shuts down:

```
The system cannot find the path specified.
C:\Users\peter\Desktop\pikaraoke-1.2\app.py:617: SyntaxWarning: invalid escape sequence '\S'
  return os.path.join(os.path.dirname(__file__), ".venv\Scripts\yt-dlp.exe")
C:\Users\peter\Desktop\pikaraoke-1.2\app.py:642: SyntaxWarning: invalid escape sequence '\p'
  legacy_directory = os.path.expanduser("~\pikaraoke\songs")
C:\Users\peter\Desktop\pikaraoke-1.2\app.py:646: SyntaxWarning: invalid escape sequence '\p'
  return "~\pikaraoke-songs"
Youtube-dl path not found! C:\Users\peter\Desktop\pikaraoke-1.2\.venv\Scripts\yt-dlp.exe
```

My first guess would be that there was an issue with my installation of ffmpeg, but when I check via command "ffmpeg -version" in the command prompt, I get this:

```
ffmpeg version 2024-07-18-git-fa5a605542-full_build-www.gyan.dev Copyright (c) 2000-2024 the FFmpeg developers
built with gcc 13.2.0 (Rev5, Built by MSYS2 project)
configuration: --enable-gpl --enable-version3 --enable-static --disable-w32threads --disable-autodetect --enable-fontconfig --enable-iconv --enable-gnutls --enable-libxml2 --enable-gmp --enable-bzlib --enable-lzma --enable-libsnappy --enable-zlib --enable-librist --enable-libsrt --enable-libssh --enable-libzmq --enable-avisynth --enable-libbluray --enable-libcaca --enable-sdl2 --enable-libaribb24 --enable-libaribcaption --enable-libdav1d --enable-libdavs2 --enable-libopenjpeg --enable-libquirc --enable-libuavs3d --enable-libxevd --enable-libzvbi --enable-libqrencode --enable-librav1e --enable-libsvtav1 --enable-libvvenc --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxavs2 --enable-libxeve --enable-libxvid --enable-libaom --enable-libjxl --enable-libvpx --enable-mediafoundation --enable-libass --enable-frei0r --enable-libfreetype --enable-libfribidi --enable-libharfbuzz --enable-liblensfun --enable-libvidstab --enable-libvmaf --enable-libzimg --enable-amf --enable-cuda-llvm --enable-cuvid --enable-dxva2 --enable-d3d11va --enable-d3d12va --enable-ffnvcodec --enable-libvpl --enable-nvdec --enable-nvenc --enable-vaapi --enable-libshaderc --enable-vulkan --enable-libplacebo --enable-opencl --enable-libcdio --enable-libgme --enable-libmodplug --enable-libopenmpt --enable-libopencore-amrwb --enable-libmp3lame --enable-libshine --enable-libtheora --enable-libtwolame --enable-libvo-amrwbenc --enable-libcodec2 --enable-libilbc --enable-libgsm --enable-libopencore-amrnb --enable-libopus --enable-libspeex --enable-libvorbis --enable-ladspa --enable-libbs2b --enable-libflite --enable-libmysofa --enable-librubberband --enable-libsoxr --enable-chromaprint
libavutil      59. 28.100 / 59. 28.100
libavcodec     61. 10.100 / 61. 10.100
libavformat    61.  5.101 / 61.  5.101
libavdevice    61.  2.100 / 61.  2.100
libavfilter    10.  2.102 / 10.  2.102
libswscale      8.  2.100 /  8.  2.100
libswresample   5.  2.100 /  5.  2.100
libpostproc    58.  2.100 / 58.  2.100
```

Any and all help would be greatly appreciated. Thank you so much for your time in advance.

**Comment by Hartebee5t** [comment]
I've yet to try a Windows installation of pikaraoke (mine's on a Pi4/5) so I'm unfamiliar with where the Windows/Linux boundary lies, but those errors look like something in the system (python maybe?) is misinterpreting those '\' Windows-style folder location delimiters as C- or Linux-style escape-sequence prefixes.
Had you typed those path definitions in a configuration file (you did mention the PATHS environment variable...), or were they generated by a script? For the app.py environment ought the folder paths be expressed in Linux-style (with forward-slash '/') characters instead of '\', maybe?

Regards
David

**Reply by StayAHDad** [reply]
Thank you for your response! The path definitions were done by a script, I believe, as I haven't really been doing anything other than starting scripts and moving files around. When I say "PATHS", I simply meant placing the ffmpeg folder into my root "C" folder and then going into WIndows' "Edit the System Environment Variables" and then adding the ffmpeg folder into the "Paths" section of my user account... Unfortunately, I'm not even sure what this is doing, and I've simply been following along with the instructions in the readme.

I do have a ".py" script called "app" which is what I assumed the app.py was referring to? Unless I'm mistaken here as well.

Sorry for the lack of any better information here. Again, I've just been following instructions like a monkey.

**Comment by vicwomg** [comment]
Seeing two issues here that look like bugs. The download path has not been
updated to work in non Unix Os.
And YouTube-dl didn’t seem to install properly. It’s possible there has
been a regression in recent versions. I’ll have a look once I get access to
a windows computer

**Reply by StayAHDad** [reply]
Oh wow - thank you so much! Really appreciate your attention on this matter :)

**Comment by vicwomg** [comment]
Ok, while I could not reproduce this issue on my windows machine, the errors all look related to python not being able to handle backslashes in the path strings, instead interpreting as escape chars. I'm not sure why some Windows configurations would have an issue with this while mine does not, but to be safe I have checked in a fix that double-escapes those path strings, replacing `\` with `\\`

I'd be curious to see if that fixes the issue. It has been merged to master. To install the fix, run

`git pull` from the pikaraoke directory. Let me know if it helps

**Reply by StayAHDad** [reply]
I'm sorry - I'm not sure if I'm doing anything wrong, but I can't seem to find anything called "git pull" from the directory. Would you know where I could find it? Thank you so much for your time on this !

**Reply by vicwomg** [reply]
Hmm, git should be installed globally.

You didn't install pikaraoke with git? Did you just download the pikaraoke source code zip file from downloads instead? If you did the latter, I'd suggest starting over: https://github.com/vicwomg/pikaraoke?tab=readme-ov-file#general-dependencies-installation

Without git, you can't pull down the latest updates

**Reply by StayAHDad** [reply]
Oh my goodness, that was my issue. I actually just noticed that I've been trying to run all of the scripts off of the zip download, despite already having used git to pull the repository onto my local machine for pikaraoke under my "user" folder.

Okay, so going through all of the steps again from scratch (including pulling latest updates via git) and running all scripts off of the proper folder under my git repository, I'm still hit with this error at the very end as I try to run pikaraoke.bat:

```
The system cannot find the path specified.
Youtube-dl path not found! C:\Users\peter\pikaraoke\.venv\Scripts\yt-dlp.exe

```
Not sure what's causing Youtube-dl to not be found by my machine, but I feel like this is the closest I've ever gotten to this working! As always, really appreciate all of your help thus far :)

**Reply by vicwomg** [reply, marked as answer]
Not sure. You ran `setup-windows.bat` first, right? 

Go the pikaraoke directory. Run `dir .venv` and `dir .venv\Scripts` to list out the contents of those directories. Does `yt-dlp.exe` show up in second file list?

**Reply by StayAHDad** [reply]
OK I JUST GOT THIS TO WORK!

I did indeed run `setup-windows.bat` prior, but noted that i didn't have `dir.venv` in my directory. This got me to try running `setup-windows.bat` again (but in command prompt so that I can see any error messages before the script closed on its own quickly) and noticed that I had an error of `'python3' is not recognized as an internal or external command, operable program or batch file.` within the setup script that runs, despite a lot of other things appearing to install fine. I did some googling and replaced "python3" in the script to simply "py" and ran the setup batch file again - and voila - I could launch the pikaraoke.bat file without issues!

I'm really sorry if all of this troubleshooting ended up simply being user error on my end, but I'm so glad to have it working now. Thank you so much for your help thus far, and for creating such a great k-karaoke system for home use :)
